api/crossref/permissions.py

```diff
@@ -0,0 +1,27 @@
+# -*- coding: utf-8 -*-
+import hmac
+import hashlib
+
+from rest_framework import permissions
+from django.core.exceptions import PermissionDenied
+
+from website import settings
+
+
+class RequestComesFromMailgun(permissions.BasePermission):
+    """Verify that request comes from Mailgun.
+    Adapted here from conferences/message.py"""
+
+    def has_permission(self, request, view):
+        data = request.data
+        signature = hmac.new(
+            key=settings.MAILGUN_API_KEY,
+            msg='{}{}'.format(
+                data['timestamp'],
+                data['token'],
+            ),
+            digestmod=hashlib.sha256,
+        ).hexdigest()
+        if signature != data.get('signature', None):
+            raise PermissionDenied('Invalid headers on incoming mail')
```

Comment:
I think this will result in a `500` (instead of the intended 403) because you're using a built-in Django error instead of a DRF error.

How about raising `rest_framework.exceptions.ParseError` instead? That will result in a 400, which I think is correct, given that we're validating the 'signature' in the request body (rather than credentials in the header).

Also, the Mailgun docs suggest using `hmac.compare_digest` to compare the hashes: https://documentation.mailgun.com/en/latest/user_manual.html#webhooks

So a more complete implementation might look like

```python
signature = hmac.new(...)
if 'signature' not in data:
    # might want to log to sentry here, since we always expect a signature
   raise ParseError('signature required in request body')
if not hmac.compare_digest(unicode(signature), unicode(data['signature'])):
    # maybe log to sentry here, too, we don't expect this to happen unless
    # somebody is doing something bad
   raise ParseError('Invalid signature')
return True
```

Reply:
Also, it'd be helpful to add a code comment linking to the [docs](https://documentation.mailgun.com/en/latest/user_manual.html#webhooks) so we know where this came from.